compiler/src/dotty/tools/dotc/interactive/Completion.scala

```diff
@@ -386,6 +365,61 @@ object Completion {
         !name.isConstructorName && name.toTermName.info.kind == SimpleNameKind
       def isStable = true
     }
+
+    object CompletionScope {
+      def from(denotations: Seq[SingleDenotation])(using Context): CompletionScope = {
+        val mappings = denotations
+          .filter(den => include(den.symbol, den.name))
+          .toList
+          .groupBy(_.name)
+          .map((name, denots) => name.stripModuleClassSuffix -> denots)
+        CompletionScope(mappings)
+      }
+
+      def fromNamed(namedDenotations: Seq[(Name, SingleDenotation)])(using Context): CompletionScope = {
+        val mappings = namedDenotations
+          .filter((name, denot) => include(denot.symbol, name))
+          .toList
+          .groupBy(_._1)
+          .map((name, namedDenots) => name.stripModuleClassSuffix -> namedDenots.map(_._2))
+        CompletionScope(mappings)
+      }
+
+      val empty = CompletionScope()
+    }
+
+    /** A wrapper over a map from names to their denotations
+     *  taking care of filtering out not matching completions
+     *  and enabling easy resolving of conflicts between different denotations with the same name
+     */
+    case class CompletionScope private(nameToDenots: Map[Name, List[SingleDenotation]] = Map.empty) {
+
+      // Merge two scopes using mappings from `that` instead of from `this` in case of name clashes
+      def mergeShadowedBy(that: CompletionScope) = CompletionScope(this.nameToDenots ++ that.nameToDenots)
+
+      // Merge two scopes but discard mappings for names that appear in both scopes
```

Comment:
```suggestion
      /** Merge two scopes but discard mappings for names that appear in both scopes */
```

Will/should a TermName `a` considered the same as a TypeName `a`?